pkg/storage/rangefeed/processor.go

```diff
@@ -0,0 +1,312 @@
+	"github.com/cockroachdb/cockroach/pkg/storage/engine"
+	"github.com/cockroachdb/cockroach/pkg/storage/engine/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+)
+
+// Processor manages a set of rangefeed registrations and handles the routing of
+// logical updates to these registrations. While routing logical updates to
+// rangefeed registrations, the processor performs two important tasks:
+// 1. it translates logical updates into rangefeed events.
+// 2. it transforms a range-level closed timestamp to a rangefeed-level resolved
+//    timestamp.
+type Processor struct {
+	clock *hlc.Clock
+	eng   engine.Engine
+	span  roachpb.RSpan
+	reg   registry
+	rts   resolvedTimestamp
+
+	regC   chan registration
+	lenC   chan int
+	eventC chan event
+	stopC  chan error
+}
+
+// event is a union of different event types that the Processor goroutine needs
+// to be informed of. It is used so that all events can be sent over the same
+// channel, which is necessary to prevent reordering.
+type event struct {
+	ops []enginepb.MVCCLogicalOp
+	ct  hlc.Timestamp
+}
+
+// NewProcessor creates a new rangefeed Processor. The corresponding goroutine
+// should be launched using the Start method.
+func NewProcessor(clock *hlc.Clock, eng engine.Engine, span roachpb.RSpan) *Processor {
+	return &Processor{
+		clock: clock,
+		eng:   eng,
+		span:  span,
+		reg:   makeRegistry(),
+		rts:   makeResolvedTimestamp(),
+
+		regC:   make(chan registration),
+		lenC:   make(chan int),
+		eventC: make(chan event, 64),
+		stopC:  make(chan error),
+	}
+}
+
+// Start launches a goroutine to process rangefeed events and send them to
+// registrations. The provided stopper is used to finish processing.
+func (p *Processor) Start(stopper *stop.Stopper) {
+	const intentPushInterval = 250 * time.Millisecond
+	const checkStreamsInterval = 100 * time.Millisecond
+
+	// WIP: annotate context.
+	ctx := context.Background()
+	stopper.RunWorker(ctx, func(ctx context.Context) {
+		// Assert that all registrations have disconnected before exiting.
+		defer func() {
+			if p.reg.Len() != 0 {
+				panic("registry not empty")
+			}
+		}()
+
+		// intentPushTicker periodically pushes all unresolved intents that are
+		// above a certain age, helping to ensure that the resolved timestamp
+		// continues to make progress.
+		intentPushTicker := time.NewTicker(intentPushInterval)
+		intentPushTickerC := intentPushTicker.C
+		defer intentPushTicker.Stop()
+		var intentPushAttemptC chan struct{}
+
+		// checkStreamsTicker periodically checks whether any streams have
+		// disconnected. If so, the registration is unregistered.
+		checkStreamsTicker := time.NewTicker(checkStreamsInterval)
+		defer checkStreamsTicker.Stop()
+
+		for {
+			select {
+			// Handle new registrations.
+			case r := <-p.regC:
+				p.reg.Register(r)
+
+			// Respond to answers about the processor goroutine state.
+			case <-p.lenC:
```

Comment:
it's nice that this is all one goroutine, but feels brittle (e.g. the way Len works is not threadsafe, correct?). maybe i'm missing something obvious, but why all the channels instead of locking?